### PR TITLE
Instant answer - Bmi calculator

### PR DESCRIPTION
**Pull request by jrenouard**
# Goodie Pull Request Template

Using this template will help us better understand your Instant Answer and assist you when necessary.  Simply copy and paste the markdown below into the description of your GitHub pull request and complete as appropriate.

**What does your Instant Answer do?**
Shows a bmi calculator when inputing "bmi"

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Community request https://duck.co/ideas/idea/4839/bmi-calculator

**What is the data source for your Instant Answer? (Provide a link if possible)**
None

**Why did you choose this data source?**
N/A

**Are there any other alternative (better) data sources?**
N/A

**What are some example queries that trigger this Instant Answer?**
bmi

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
People trying to get in shape

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
Yes it is https://duck.co/ideas/idea/4839/bmi-calculator

**Which existing Instant Answers will this one supercede/overlap with?**
None (I hope)

**Are you having any problems? Do you need our help with anything?**
I'm not sure what the text version should return.

**Where did you hear about DuckDuckHack? (For first time contributors)**
On your job listing page  https://duck.co/help/company/hiring

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
screen shot 2015-02-03 at 19 40 06
screen shot 2015-02-03 at 19 39 57
screen shot 2015-02-03 at 19 40 12

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
[X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 7
        [] IE 8

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [X] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```

### CODE REVIEW

File: lib/DDG/Goodie/Bmi.pm [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+zci is_cached   => 1;
+
+name "Bmi";
+description "Pops a bmi calculator";
+primary_example_queries "bmi";
+category "calculations";
+topics "special_interest", "everyday";
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Bmi.pm";
+attribution github => ["jrenouard", "Julien Renouard"],
+            twitter => "@jurenouard";
```
Comment (mintsoft) [inline lib/DDG/Goodie/Bmi.pm]: This is usually just "jurenouard"

File: share/goodie/bmi/bmi.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+null
```
Comment (mintsoft) [inline share/goodie/bmi/bmi.txt]: I presume this file should be populated or empty?

**Comment by mintsoft** [comment]
@jrenouard Thanks for this! It's different from all the other Goodies as it's using JS, usually we use spices for these.

@moollaza @jagtalon do we switch this to a spice or is it slotting into the new templating system?

**Comment by moollaza** [comment]
Hmm. @mintsoft my initial reaction is that this should be moved to Spice to allow for templates and JS. Let me speak to the team about how close we are to supporting the Goodie Templates..

File: share/goodie/bmi/bmi.txt (hunk 2 above)
Comment (jrenouard) [inline share/goodie/bmi/bmi.txt]: I have no idea what to return for the text version because the calculator is html/js, please advise. Thanks

**Comment by jrenouard** [comment]
@moollaza @mintsoft just let me know how I can help.

**Comment by moollaza** [comment]
@jrenouard apologies, you didn't misinterpret the chart, in general the Goodie route would have worked, but Spice (because they deal with callbacks) are typically our go-to IA type when dealing with JavaScript and frontend interaction like this IA has.

There are two options here:

1. Convert to a Spice and use templates like the Timer and Stopwatch IA's
2. Wait until the Goodie Templates are available. Currently we're working to bring all the Spice templating to Goodies so that they too can produce a templated result which opens them up to much easier implementations and allows for a better way to provide JavaScript for frontend interaction. This could even be implemented as a pure JS Instant Answer (no Perl backend) which is something we're considering, but don't have concrete plans for yet.

Option 1 means we can move this along now and have people use it. Option 2 would be really cool as this could be one of the first templated Goodies :)

What do you guys think?

**Comment by mintsoft** [comment]
I'm not fussed totally up to @jrenouard

**Comment by jrenouard** [comment]
I'm always in for the cool option, so Option 2 it'll be.
What are your thoughts on adding CoffeeScript?

**Comment by moollaza** [comment]
>I'm always in for the cool option, so Option 2 it'll be.
Great, I've marked this PR as "On Hold" -- We can pick this up again when the templates are available.

>What are your thoughts on adding CoffeeScript?
For now, we're just sticking to JavaScript. It's what we use everywhere else and what our team (and contributors) are most familiar with.

**Comment by moollaza** [comment]
@jrenouard so you know, the next version of Goodie templates is unofficially ready. I'd like to take a stab at implementing this to see how it fares and show it to the team. If it's good I'll send the code your way or just merge this and update it myself.

Just wanted to keep you in the loop!

**Comment by jrenouard** [comment]
Hey @moollaza, thanks for keeping me updated.
It definitely relies heavily on JS, I'll be glad to help you test your new templating system, let me know when we can do that.

**Comment by moollaza** [comment]
@jrenouard just wanted to update you. We basically have everything figured out for Goodies to use custom templates and execute JS. Examples and documentation are coming, the new 2048 Game Pull Request is likely going to be one of the first to use JS and likely this one as well :)

**Comment by jrenouard** [comment]
@moollaza That's great news, let me know when I can rewrite this using your templates system.

**Comment by moollaza** [comment]
@jrenouard you can now move forward with this IA! Templates and JS are ready for Goodies :)

There's a few Goodies in the repo that use custom templates and JS but I've made a starter IA for you here as an example: https://github.com/duckduckgo/zeroclickinfo-goodies/commit/04e2a70136aca039edb6a30b10d1f8f36907c868

You can look at the BPMToMS, Minecraft and CheatSheets Goodies for more examples. The 2048 Game PR is also another example of a tempalted Goodie with a handlebars template and JS.

We don't have docs written yet for this but the idea is nearly identical to the way Spice templates are handled and documented.

Let me know if you have any questions!

**Comment by moollaza** [comment]
Ping! @jrenouard are you interested in finishing this up? We'd love to see it go live!

**Comment by jrenouard** [comment]
Hey @moollaza I'm really interested in finishing this, I've had a lot of work during the past few weeks, but I should be able to take a look at this next week.

**Comment by moollaza** [comment]
@jrenouard glad to hear you're still interested! There's no rush on getting it done, I was just checking in to see if you're still around, otherwise I was going to ask the community to finish it up.

I'm looking forward to seeing this go live!

**Comment by javathunderman** [comment]
@jrenouard ping! Need any help?

**Comment by jrenouard** [comment]
Hey @moollaza and @javathunderman 
Sorry about the delay.
I'm not familiar with handlebar so I just moved my JS code into the onShow.
I'm not sure if this is the way you want it to be. Let me know.

**Comment by moollaza** [comment]
@jrenouard sorry for the delay!

This is on the right path but not quite there.

The idea is that you should be doing as much setup and pre-work in the `build` function, instead of the `onShow`. You have access to the `ops` object which contains the `structured_answer` returned from the Perl.

The flow should be like so:

Perl handles triggering, query verification and as much grunt work as you want. It builds your `structured_answer` and ideally sets up your `data` with whatever info you need to be passed to the frontend. It also specifies your templates and other info (i.e. `group => "text"`, `content => 'DDH.Bmi.content'`). Your data should specify the `title` and `subtitle` (or `subtitle_content`, a custom handlebars file).

You have a `content.handlebars` file in `/share/Goodie/bmi` which is a handlebars file (HTML + variables). Which contains the HTML markup for the `<form>`

Your JS file defines some functions for calculating BMI and applies an even handler to the button in your UI.

Take another look at the demo I built, https://github.com/duckduckgo/zeroclickinfo-goodies/commit/04e2a70136aca039edb6a30b10d1f8f36907c868 and play around with it to see how it all works.

You'll want to move all the static functions out of `onShow`. The `onShow` should be used to initially attach an event listener to the "Calculate" button using jQuery.

Let me know if this make more sense!

**Comment by mintsoft** [comment]
@jrenouard just checking in! Are you still interested in finishing this up?

**Comment by jrenouard** [comment]
@mintsoft please review https://github.com/duckduckgo/zeroclickinfo-goodies/commit/ae2e3d655a1efe20b0950eda8b841a3408fd8e51

**Comment by mintsoft** [comment]
@jrenouard looks good on the whole; just needs the markup breaking out. If you look at the example that moollaza provided, handlebars is used for the templating: https://github.com/duckduckgo/zeroclickinfo-goodies/commit/04e2a70136aca039edb6a30b10d1f8f36907c868

**Comment by jrenouard** [comment]
@mintsoft like this https://github.com/jrenouard/zeroclickinfo-goodies/commit/ea005f2b82c52571006e2a024f55b340955a9681 ?

**Comment by mintsoft** [comment]
@jrenouard Basically yes! There's a little more markup in the JS

**Comment by moollaza** [comment]
ping! @jrenouard how's it going? Do you need help with anything here?

Looking at the code I think you may not fully understand how this should all work.

To reiterate the flow is like so:

1. IA Triggers
2. Goodie executes `handle()` which can investigate the query and prepare data if needed
    - If query is "bmi calculator" the Goodie likely won't need to return anything in it's `data` hash, but it should still specify the template, id, name, source etc
    - If the query has info we need, like "bmi 6ft tall, 200lbs" then you'll want to pass that along to the frontend via your `data` object.
    - You could even calc the BMI and pass it along, but we'll need JS logic for calculating the BMI from the form inputs anyway, so it seems easiest to pass along the data
3. The hash `return`'ed from the Perl get's converted to JSON and passed along to your JavaScript `build() function:
    - Here you will setup your JS logic and handlers for reacting to the form and creating output when it's clicked.
    - You'll also want to take a look at the JSON object we passed along to see if it has any data from the Perl. If it does, you'll probably want to use this to display a result on the screen immediately. If not, it means we'll need to show a blank form and wait for the user to input something and click the button
4. Your JS will ultimately return an object, which is used to `$.extend()` the JSON hash we got from the Perl. This helps you add to the original hash, or overwrite it. For example, if you needed to change the template you could do so from the JS.

Does this make sense?

**Comment by jrenouard** [comment]
hey @moollaza thanks for the detailed answer but I won't have time to finish this in the near future.
Please ask the community, and sorry about the time wasted...

**Comment by moollaza** [comment]
>hey @moollaza thanks for the detailed answer but I won't have time to finish this in the near future.
Please ask the community, and sorry about the time wasted...

@jrenouard not a problem! Thanks for working with us so far! We really hope you'll stick around and continue to contribute (when you have the time :smiley: )

**Comment by javathunderman** [comment]
I'd be happy to pick this up for Saturday's Q&H!

**Pull request closed without merging**